### PR TITLE
Clean Up ZeRO

### PR DESCRIPTION
**Pull request by awgu**
**Overview:**
Being relatively new to PyTorch and ZeRO, I found parts of the code slightly hard to follow. This change strives to clean up the `ZeroRedundancyOptimizer` code in `zero_redundancy_optimizer.py` by reorganizing some computations, making variable names more explicit and consistent, and unifying terminology in the documentation. The goal is for the code to be easier to extend afterwards.

**Changes:**
1) `state_dict()`: The [logic](https://github.com/pytorch/pytorch/blob/85517a2b700a5abc0b38f53ce8c99404cd67db79/torch/distributed/optim/zero_redundancy_optimizer.py#L510) for updating the global `state_dict` with each rank's local `state_dict` is simplified and made more explicit. Notably, the `dict` [`local_index_to_param_id`](https://github.com/pytorch/pytorch/blob/85517a2b700a5abc0b38f53ce8c99404cd67db79/torch/distributed/optim/zero_redundancy_optimizer.py#L513) is unneeded. It maps `local_pg["params"][i]` to `id(global_pg["params"][i])`, so it is equivalent to make a single pass over both lists in tandem, effectively iterating over `i`, without a need for the explicit `dict`.
2) `_update_trainable()`: The function [initializes](https://github.com/pytorch/pytorch/blob/85517a2b700a5abc0b38f53ce8c99404cd67db79/torch/distributed/optim/zero_redundancy_optimizer.py#L597) the local optimizer if it does not exist. I am unaware of any reason for the local optimizer to be destroyed after initialization, so I moved that logic to its own function `_init_local_optimizer()`, which is called once in the constructor. 
After [discussion](https://github.com/pytorch/pytorch/pull/60285#discussion_r654706728), I removed the function `_update_trainable()` itself in favor of adding a check for `parameters_as_bucket_view` in `build_param_buckets()` directly. 
3) `rank_local_state_dict()`: This [function](https://github.com/pytorch/pytorch/blob/85517a2b700a5abc0b38f53ce8c99404cd67db79/torch/distributed/optim/zero_redundancy_optimizer.py#L528) is currently broken. It appears to be legacy and relies on the input `state_dict` to have the key `"partitions"`. For now, I have removed it and added an [issue](https://github.com/pytorch/pytorch/issues/60284). Is it a notable use case to want to access another rank's `state_dict` in particular (as opposed to consolidating the entire state and then accessing)? 
4) `local_state_dict():` After [discussion](https://github.com/pytorch/pytorch/pull/60285#discussion_r655571043), I removed the function. 
5) `partition_parameters()`: After [discussion](https://github.com/pytorch/pytorch/pull/60285#discussion_r654708183), I renamed the function to `_partition_parameters()` to mark it as private.
6) `_param_to_index`: After [discussion](https://github.com/pytorch/pytorch/pull/60285#discussion_r654828100), I changed the key to be the parameter itself rather than its integer ID.
7) `buckets`: I renamed the data structure to `_buckets` to mark it as private.
8) Terminology: I tried to reduce the set of terms being used instead of juggling a number of synonyms. In particular, I made an effort to distinguish between "local" and "global" and to make names more indicative of typing.
9) Style: Per the [PyTorch contributing guide](https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#writing-documentation), I made all docstrings abide by the 80 character limit, except for the one [line](https://github.com/andwgu/pytorch/blob/554891f6faa764c76dec4afb1107cb5aa88ef589/torch/distributed/optim/zero_redundancy_optimizer.py#L142) showing the example ZeRO usage. Some code lines violate the limit for readability. Also, I unified some of the minor stylistic usages out of habit.

**Test Plan:**
The test suite passes as expected (on the AI AWS cluster):
```
gpurun python test/distributed/optim/test_zero_redundancy_optimizer.py
```
I visually inspected the generated HTML doc (as generated following [this](https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#writing-documentation)).

### CODE REVIEW

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 1]
```diff
@@ -44,76 +44,94 @@ def _recursive_copy_to_device(value: Any, non_blocking: bool, device: torch.devi
...
     return param.requires_grad
 
 
-def _broadcast_object(
-    obj: Any,
-    src_rank: int,
-    group: object = dist.group.WORLD,
-    dist_device: torch.device = torch.device("cpu"),
-) -> Any:
+def _broadcast_object(obj: Any, src_rank: int,
```
Comment (blefaudeux) [inline torch/distributed/optim/zero_redundancy_optimizer.py:54]: this is not needed anymore with modern pytorch, it was there initially because this functionality was not present, I think that you could remove it altogether

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 2]
```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
...
             # Either replace the existing bucket or create it
-            if len(self.buckets) != rank:
-                self.buckets[rank] = bucket
+            if len(self._buckets) != rank:
+                self._buckets[rank] = bucket
             else:
-                self.buckets.append(bucket)
+                self._buckets.append(bucket)
 
     def _update_trainable(self) -> None:
```
Comment (pritamdamania87) [inline torch/distributed/optim/zero_redundancy_optimizer.py:623]: I think we can get rid of this method entirely and just call `_build_param_buckets` and inside `_build_param_buckets` have the check for `self.parameters_as_bucket_view`.

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 3]
```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
...
-        # - get the pytorch compliant parameter indexing
+        # Get the possibly-stale global optimizer state that uses global
+        # parameter indexing
         state_dict = super().state_dict()
 
-        # - go through the per-shard states, which are all indexed locally
-        for rank, s in enumerate(self._all_states):
-            # -- match the local indexing and the global partition, update the corresponding saved state globally
-            for local_pg, global_pg in zip(s["param_groups"], self.partition_parameters()[rank]):
-                local_index_to_param_id = {
```
Comment (blefaudeux) [inline torch/distributed/optim/zero_redundancy_optimizer.py:496]: this was for compatibility with pytorch 1.5 (this comes from fairscale, which is compatible with older pytorch, current-2 and at the time this came out it was 1.5). this is not needed anymore  indeed. There were a couple of update PRs for this code which never got reviewed, it's really good that you can spend some time in that

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 4]
```diff
@@ -194,133 +218,134 @@ def _clear_cache(self) -> None:
...
-
-        # NOTE: In practice, `broadcast` is used, which is wasteful (gather would have been appropriate)
-        # compatibility issues with some backends make the use of broadcast mandatory for now.
-        # a possible follow up would be to move all sharded state management to RPC RRef
+        # Pull the sharded state from all ranks and store them in rank order
+        empty_messenger = torch.tensor([0], dtype=torch.uint8, device=self._default_device)
 
-        self._all_states = []
+        # NOTE: We wastefully use `broadcast()` (e.g. instead of `gather()`)
+        # due to compatibility issues with some backends; a possible follow-up
```
Comment (pritamdamania87) [inline torch/distributed/optim/zero_redundancy_optimizer.py:281]: I'm guessing this is referring to NCCL? If so we should mention this here and if we don't already have a gh issue, we should mention the lack of gather support in NCCL and add a code pointer for this to that issue.

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 5]
```diff
@@ -194,133 +218,134 @@ def _clear_cache(self) -> None:
...
+            A :class:`list` of ``param_groups`` (which is a :class:`list` of
+            :class:`dict`) where each element of the list contains the
+            ``param_groups`` for a rank. Element 0 corresponds to rank 0, etc.
+            Each rank stores the ``param_groups`` for all of the ranks for the
+            collective communication in :meth:`step`.
+
+        NOTE: ``test_sharding()`` and ``test_add_param_group()`` in
+        ``test_zero_redundancy_optimizer.py`` rely on this function using the
+        sorted-greedy algorithm for partitioning. If the algorithm is changed,
+        please re-examine those two tests.
```
Comment (pritamdamania87) [inline torch/distributed/optim/zero_redundancy_optimizer.py:338]: We shouldn't have notes about test methods in the docs of public methods. Also, shouldn't this method be private? Do we actually expect users to call this method? If not, it should be private.

**Comment by blefaudeux** [comment]
looks good to me, some of these changes were overdue indeed. Some of the weird parts that you're seeing came from the retrocompatibility with pytorch 1.5, which notably had a different way of indexing the parameters in the state dict (and this code, coming from fairscale, had to be compatible with 1.5, 1.6 and 1.7 checkpoints). In general this was not kept in sync with Fairscale, and that's a shame. I would recommend that you have a look at the unit tests (fairscale's are [here](https://github.com/facebookresearch/fairscale/blob/master/tests/optim/test_oss.py), not perfect either but there are probably a couple more cases being checked), they could probably be improved also, from experience it's kind of easy to overlook a corner case

**Comment by andwgu** [comment]
> looks good to me, some of these changes were overdue indeed. Some of the weird parts that you're seeing came from the retrocompatibility with pytorch 1.5, which notably had a different way of indexing the parameters in the state dict (and this code, coming from fairscale, had to be compatible with 1.5, 1.6 and 1.7 checkpoints). In general this was not kept in sync with Fairscale, and that's a shame. I would recommend that you have a look at the unit tests (fairscale's are [here](https://github.com/facebookresearch/fairscale/blob/master/tests/optim/test_oss.py), not perfect either but there are probably a couple more cases being checked), they could probably be improved also, from experience it's kind of easy to overlook a corner case

For sure, thanks for taking a look! I will read the Fairscale tests.

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 6]
```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
...
-            If the state has not been consolidated, this returns a shard's worth,
-            not the global state.
-
-        .. warning:
-            Returning the global state is limited to the replica which was
-            responsible for the consolidation. The state may also not be up to
-            date, depending on when :meth:`consolidate_state_dict` was last called.
+            If the state has not been consolidated to this rank, this raises a
+            runtime error, and even if it has, the state may not be up-to-date,
+            depending on when :meth:`consolidate_state_dict` was last called.
```
Comment (pritamdamania87) [inline torch/distributed/optim/zero_redundancy_optimizer.py:506]: I'm wondering if we need this interaction between state_dict and consolidate_state_dict? Every time we call `.step()` can we update a variable `needs_consolidation=True`. Then when state_dict is called, we first consolidate the state dicts if `needs_consolidateion=True`, cache the result and set `needs_consolidation=False`. This way the user only has to worry about calling `state_dict`.

**Review by pritamdamania87 — COMMENTED** [review]
Thanks for cleaning up the implementation!

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 7]
```diff
@@ -147,44 +165,50 @@ def __init__(
...
+        self._optim_defaults = defaults
         self._optim_constructor = optimizer_class
+        self._init_local_optimizer()
 
-        # Optional consolidated optimizer state
-        self._all_states: List[Dict[str, Any]] = []
+        self.parameters_as_bucket_view = parameters_as_bucket_view
+        self._is_trainable_mask = self._get_is_trainable_mask()
+        self._buckets: List[torch.Tensor] = []
+        self._update_trainable()
```
[the remark this answers is not in the document]
Comment (andwgu) [reply torch/distributed/optim/zero_redundancy_optimizer.py:206]: I think following your other comment and removing `_update_trainable()` altogether in favor of only `_build_param_buckets()` (which checks `parameters_as_bucket_view`) might make this more clear. Then, the constructor will call `_init_local_optimizer()` to initialize the optimizer and call `_build_param_buckets()` to build the buckets, as the function names suggest.

**Comment by mrshenli** [comment]
> I visually inspected the generated HTML doc (as generated following this). Only the "public" functions (i.e. those without any preceding _) are rendered, so I am not sure how to verify the other functions.

Functions start with `_` are consider private functions, and we won't expose them to API docs. Their docstring can still help future developers to understand the code.

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 8]
```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
...
-                        global_id = self._param_to_index[local_index_to_param_id[local_param_index]]
-                        state_dict["state"][global_id] = s["state"][local_param_index]
-
-        # Make sure that the parameters are sorted in the state, as expected
+        # Update the global optimizer state with local state information,
+        # factoring in the translation from local to global indexing
+        for rank, local_state_dict in enumerate(self._all_state_dicts):
+            local_param_groups = local_state_dict["param_groups"]
+            global_param_groups = self.partition_parameters()[rank]
+            assert len(local_param_groups) == len(global_param_groups)
```
Comment (mrshenli) [inline torch/distributed/optim/zero_redundancy_optimizer.py:525]: add an error message for this assert?

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 9]
```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
...
+                self._buckets[rank] = bucket
             else:
-                self.buckets.append(bucket)
+                self._buckets.append(bucket)
 
     def _update_trainable(self) -> None:
         r"""
         Updates the partitioning and communication patterns if the trainability
-        (``requires_grad``) of some parameters changed.
+        (``requires_grad``) of some parameters has changed.
```
Comment (mrshenli) [inline torch/distributed/optim/zero_redundancy_optimizer.py:626]: has -> have?

**Review by mrshenli — COMMENTED** [review]
Thanks a lot for improving code quality!

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 1 above)
Comment (andwgu) [reply torch/distributed/optim/zero_redundancy_optimizer.py:54]: I looked into replacing `_broadcast_object()` with `torch.distributed.broadcast_object_list()` as [suggested](https://github.com/pytorch/pytorch/pull/60285/files/554891f6faa764c76dec4afb1107cb5aa88ef589#r654825551) by @mrshenli. However, `broadcast_object_list()` yields a significant slowdown in my preliminary testing. Precisely, running `test_collect_shards()` on the AI AWS cluster with a world size of 4 regresses from ~6 seconds to ~18 seconds (where the affected function is `consolidate_state_dict()`).

I am unable to resolve the issue, but this is what I found: There are two (possibly-related) points of slowdown in `broadcast_object_list()`:
1) The implicit `torch.cuda.synchronize()` resulting from [moving data from `object_tensor_sizes` from GPU to CPU via `.item()`](https://github.com/pytorch/pytorch/blob/5824a866b72c251ad47a9c16dc652e49cfd7e234/torch/distributed/distributed_c10d.py#L1781) takes ~4 seconds for one of the broadcasts on ranks 1 and 2 (while generally it takes ~1 millisecond). This confuses me since `_broadcast_object()` follows the exact same data movement pattern, yet it does not suffer from a slow synchronization.
2) Unlike `_broadcast_object()`, `broadcast_object_list()` pickles/unpickles the object via [`_object_to_tensor()`](https://github.com/pytorch/pytorch/blob/5824a866b72c251ad47a9c16dc652e49cfd7e234/torch/distributed/distributed_c10d.py#L1757)/[`_tensor_to_object()`](https://github.com/pytorch/pytorch/blob/5824a866b72c251ad47a9c16dc652e49cfd7e234/torch/distributed/distributed_c10d.py#L1797), respectively. `_tensor_to_object()` consistently takes ~4 seconds. 
I am unable to see any connection between the two, but the similarity between their latencies is curious. Together, these two points cause `consolidate_state_dict()` to take >4 seconds when using `broadcast_object_list()`, while it only takes a few milliseconds when using `_broadcast_object()`. 

Any help investigating this is greatly appreciated!

NB: Without pickling, the tensor allocations are 939, 939, and 1131 bytes, while with pickling, the allocations are 571, 591, and 755 bytes.

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 10]
```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
...
-            * state - a dict holding current optimization state. Its content
-                differs between optimizer classes.
-            * param_groups - a dict containing all parameter groups
+            The state of the local optimizer as a :class:`dict`, containing
+            two entries:
+
+            * state
+                a dict holding the current optimization state, where the exact
+                contents depends on the optimizer class.
+            * param_groups
```
[the remark this answers is not in the document]
Comment (andwgu) [reply torch/distributed/optim/zero_redundancy_optimizer.py:492]: Screen Shot 2021-06-20 at 9 02 40 PM
I changed it to be like this because any other text on the same line as `* state` or `* param_groups` appears in the grey box, but anything in the next line appears underneath in the white. The old way looked strange with `differs between optimizer classes.` being indented and in the white, while `* state - a dict holding current optimization state. Its content` was in the grey box above.

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 9 above)
Comment (andwgu) [reply torch/distributed/optim/zero_redundancy_optimizer.py:626]: Very minor, but I believe the current way is correct since it should be `trainability` [...] `has changed` rather than `trainability` [...] `have changed` (`trainability` being singular not plural)
:)

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 4 above)
Comment (blefaudeux) [reply torch/distributed/optim/zero_redundancy_optimizer.py:281]: from the top of my head that was with older Gloo (compatibility matrix was pytorch 1.5/1.6/1.7 & NCCL/Gloo & CPU/GPU  with gloo), maybe that this is worth revisiting ?

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 6 above)
Comment (blefaudeux) [reply torch/distributed/optim/zero_redundancy_optimizer.py:506]: No, that would not work because of the collective needs. The issue here is that users might call `.state_dict()` on one rank only (in practice they often will), which means that you cannot consolidate at that point with any of the collective communication primitives. Something like RPC would alleviate that, but that's a bigger change, without this you need the collective consolidation before you can checkout (and doing this for every step is super slow, of course).

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 11]
```diff
@@ -346,16 +371,19 @@ def _param_to_rank(self) -> Dict[torch.Tensor, int]:
...
+        optimizer state.
+
+        NOTE: This assumes that the global optimizer state's indexing (in
+        ``state_dict``) follows a linear ordering over the parameter groups.
         """
-        if len(self._param_to_index_cache) == 0:
-            self._param_to_index_cache = {
+        if len(self._param_id_to_index_cache) == 0:
+            self._param_id_to_index_cache = {
                 id(p): i for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))
```
[the remark this answers is not in the document]
Comment (blefaudeux) [reply torch/distributed/optim/zero_redundancy_optimizer.py:384]: (seconding that personally, but soft preference)

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 12]
```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
...
                 self.optim.state[param] = _recursive_copy_to_device(value, non_blocking=True, device=param.device)
 
         super().load_state_dict(state_dict)
 
-        # Sync with the optimizer param groups
+        # Sync the input state with the exposed and local optimizer states
         self._sync_param_groups(state_dict["param_groups"], self.param_groups)
         self._sync_param_groups(self.param_groups, self.optim.param_groups)
 
     def local_state_dict(self) -> Dict:
```
Comment (blefaudeux) [inline torch/distributed/optim/zero_redundancy_optimizer.py:481]: I would remove that method altogether, I don't think that there's a use case for it actually (we don't expose a way to reload local state dicts)

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 13]
```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
...
-        r"""
-        Make all params which are on the same device and tied to the same rank
-        views of a single buffer. This is used at construction time, and
-        anytime parameter trainability is changed (frozen or unfrozen) and
-        ``_update_trainable`` is called.
+        NOTE: The current implementation assumes that each rank stores all of
+        its parameters (i.e. ``self._all_params``) on a single device. This
+        means that there should be exactly ``world_size``-many buckets.
+
+        NOTE: The current implementation assumes that all of the parameters in
```
Comment (blefaudeux) [inline torch/distributed/optim/zero_redundancy_optimizer.py:582]: could be worth it adding an assert with a nice error message ? not super likely to be triggered, but in case that happens it would help debugging a fair bit

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (andwgu) [reply torch/distributed/optim/zero_redundancy_optimizer.py:384]: I examined the hashing for both `Tensor` and `Parameter` and neither seem to use `id()` by default.
```
>>> model = torch.nn.Linear(1, 1)
>>> t = model.weight.data  # t: torch.Tensor
>>> d = dict()
>>> d[t] = 1
>>> d[id(t)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 139809367343568
>>> p = model.weight  # p: torch.nn.parameter.Parameter
>>> d = dict()
>>> d[p] = 1
>>> d[id(p)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 139809367327664
```
Let me know if I tested this incorrectly.

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 13 above)
Comment (andwgu) [reply torch/distributed/optim/zero_redundancy_optimizer.py:582]: The assumptions in the `NOTE: ...`s should be verified at the very beginning of the constructor (line 171). I was intending the notes for documentation purposes (for a user reading the source or for a developer planning to modify the implementation). 

Do you think we should do a second pass of verifications each time this function `_build_param_buckets()` is called?

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mrshenli) [reply torch/distributed/optim/zero_redundancy_optimizer.py:384]: Thanks for checking this. My mistake.

[the remark this answers is not in the document]
Comment (mrshenli) [reply torch/distributed/optim/zero_redundancy_optimizer.py:384]: Hold on, I am not sure the test is correct. We are passing a `Tensor: int` pair, but checking an `int: int` pair? Isn't that suppose to fail? Below is Tensor'e hash function:

https://github.com/pytorch/pytorch/blob/8dd1dc89cbe1154640aa058b1eb09682144cfedd/torch/_tensor.py#L603-L606

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 9 above)
Comment (mrshenli) [reply torch/distributed/optim/zero_redundancy_optimizer.py:626]: Oh I see, my bad, thanks!

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (andwgu) [reply torch/distributed/optim/zero_redundancy_optimizer.py:384]: Ah, I misunderstood. Yes, I believe we can switch back to `_param_to_index` and make it of type `dict[torch.Tensor, int]`. Then, instead of the key being `id(p)`, it can directly be `p`.

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 1 above)
Comment (andwgu) [reply torch/distributed/optim/zero_redundancy_optimizer.py:54]: I think I found the issue, and fixing this will require changing code in `distributed_c10d.py`. Here is my hypothesis:

Currently, `broadcast_object_list()` takes care to specially handle the NCCL backend case, making sure to move `object_sizes_tensor` and `object_tensor` to device memory before broadcasting. Upon receiving the broadcasted tensor, the code calls `_tensor_to_object()`, which loads the tensor to an object. The existing implementation uses a `Pickler` and `Unpickler` to save/load, but I have verified that the slowdown behavior is replicated using `torch.save()` and `torch.load()`, which itself[ uses pickling](https://pytorch.org/docs/stable/_modules/torch/serialization.html#load). For the remainder, I will assume a `torch.save()`/`torch.load()`-based implementation of `_object_to_tensor()` and `_tensor_to_object()`.

The crux is the default behavior of `torch.load()` (and presumably `Unpickler.load()`), which deserializes (on CPU) and loads to the device **where the tensor was saved**. Suppose that the world size is 4. Then, in the received `local_state_dict` on rank `j` from the broadcast, the contained tensors are loaded to `cuda:i` where `i != j`. The time bottleneck appears to be exactly in loading these received objects to device, taking `> 3.5` seconds. 

The bandaid fix I have right now is changing:
* `_object_to_tensor(obj)` to `_object_to_tensor(obj, pickle=True)`
* `_tensor_to_object(tensor, tensor_size)` to `_tensor_to_object(tensor, tensor_size, pickle=True, device=None)`

If `pickle == True`, then the behavior is as-is to preserve backward compatibility. Otherwise, `torch.save(obj, f)` and `torch.load(io.BytesIO(buf), map_location=device)` are used instead of the `_pickler` and `_unpickler`, respectively. Then, `broadcast_object_list()` passes in `device=current_device` to `_tensor_to_object()` if using NCCL backend. I have confirmed that this change removes the slowdown.

However, this solution is only a temporary fix, and I think a more thorough examination is needed. I would need to see if a solution is feasible while still using the `Pickler`/`Unpickler` (though at first glance this is not promising since the `Unpickler.load()` provides no options).

I do wonder how much this behavior has affected the speed of other use cases. It seems to be pretty common that the received tensor from the broadcast is from a different device -- in fact, that should be the typical case?

Comment (mrshenli) [reply torch/distributed/optim/zero_redundancy_optimizer.py]: Discussed with @andwgu offline on this. Since there are gaps in existing `broadcast_object_list()`, we are going to keep the current `_broadcast_object` method. Later, if bandwidth allows, we can work on a followup PR to improve `broadcast_object_list()` and use it here.

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 14]
```diff
@@ -9,16 +9,19 @@
...
 from torch.optim import Optimizer
-import logging
 
 __all__ = ["ZeroRedundancyOptimizer"]
 
 
 # Credits:  classy_vision/generic/distributed_util.py
-def _recursive_copy_to_device(value: Any, non_blocking: bool, device: torch.device) -> Any:
+def _recursive_copy_to_device(
+    value: Any, non_blocking: bool,
```
Comment (mrshenli) [inline torch/distributed/optim/zero_redundancy_optimizer.py:22]: shall we use one arg per line format to stay consistent with other APIs in this file and other files?

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 15]
```diff
@@ -9,16 +9,19 @@
...
 import torch.distributed as dist
 from torch.optim import Optimizer
-import logging
 
 __all__ = ["ZeroRedundancyOptimizer"]
 
 
 # Credits:  classy_vision/generic/distributed_util.py
-def _recursive_copy_to_device(value: Any, non_blocking: bool, device: torch.device) -> Any:
+def _recursive_copy_to_device(
```
Comment (mrshenli) [inline torch/distributed/optim/zero_redundancy_optimizer.py:21]: (no need to change this PR, just a heads-up) there is another `recursive_to` in DDP. It might be helpful to consolidate these two into one common utility function. 

https://github.com/pytorch/pytorch/blob/4887c6e401b73c6b024a344076899fd8cdf4a26b/torch/nn/parallel/distributed.py#L907

File: torch/distributed/optim/zero_redundancy_optimizer.py [hunk 16]
```diff
@@ -419,172 +450,165 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
...
-    def _sync_param_groups(source: List[Dict[Any, Any]], destination: List[Dict[Any, Any]]) -> None:
-        r"""Sync learning rate and other optimizer attributes (needed to support schedulers)."""
+        This function is called in the constructor and any time parameter
+        trainability is changed.
 
-        for source_group, destination_group in zip(source, destination):
-            # Sync everything but the parameters
-            for k in filter(lambda x: x != "params", source_group.keys()):
-                destination_group[k] = source_group[k]
+        NOTE: The current implementation assumes that each rank stores all of
```
Comment (mrshenli) [inline torch/distributed/optim/zero_redundancy_optimizer.py:566]: (no need to change this PR, just FYI) if you would like to add a note block, this needs to be `.. note::`. But since this is a private API and the docs won't render, the current one also works. 

https://github.com/pytorch/pytorch/blob/4887c6e401b73c6b024a344076899fd8cdf4a26b/torch/nn/parallel/distributed.py#L216

**Review by mrshenli — APPROVED** [review]
LGTM!

@andwgu you might want to rebase to avoid the merge conflicts in tests.

File: torch/distributed/optim/zero_redundancy_optimizer.py (hunk 13 above)
Comment (blefaudeux) [reply torch/distributed/optim/zero_redundancy_optimizer.py:571]: sorry, a bit late to react, I meant checking for this assumption explicitly (at least the first time the buckets are built ?) and nicely error out if any issue, but maybe that it was already there indeed. Non blocking, just a brain dump
